Add madvr envy integration

iloveicedgreentea opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This adds an integration for the madvr envy http://madvrenvy.com

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/33401/

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

iloveicedgreentea committed:
feat: Add madvr envy

iloveicedgreentea committed:
fix: await and pass entry directly

iloveicedgreentea committed:
fix: add attributes and unique id for sensors

iloveicedgreentea committed:
fix: reflect power state well, improve state detection

iloveicedgreentea committed:
fix: don't connect on init, add options, add reload on change, keep on during test

iloveicedgreentea committed:
fix: cancel tasks on unload

iloveicedgreentea committed:
fix: test connection via library

iloveicedgreentea committed:
fix: wait for boot time

iloveicedgreentea committed:
docs: add readme and license

iloveicedgreentea committed:
fix: broken pipe in lib

iloveicedgreentea committed:
fix: detect out of band power off

iloveicedgreentea committed:
fix: improve extra attributes

iloveicedgreentea committed:
fix: fix unloading, add config flow test, limit to one platform

joostlek commented on homeassistant/components/madvr/Readme.md:
This is not common so I think its best to keep this in the docs

iloveicedgreentea replied:
okay thank you for the prompt review, will fix these

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,69 @@
+from .coordinator import MadVRCoordinator
+from .utils import cancel_tasks
+
+# For your initial PR, limit it to 1 platform.
+PLATFORMS: list[Platform] = [Platform.REMOTE]
+
+# Alias name should be prefixed by integration name
```
Comments dont add much

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,69 @@
+PLATFORMS: list[Platform] = [Platform.REMOTE]
+
+# Alias name should be prefixed by integration name
+type MadVRConfigEntry = ConfigEntry[MadVRCoordinator]  # noqa: F821
```
Why the noqa

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,69 @@
+    """Set up the integration from a config entry."""
+    name = entry.data["name"]
+    madVRClient = Madvr(
+        host=entry.data["host"],
+        logger=_LOGGER,
+        port=entry.data.get("port", 44077),
```
Please use constants `CONF_*`

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,69 @@
+        mac=entry.data.get("mac", ""),
+        name=name,
+    )
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+    hass.data[DOMAIN]["entry_id"] = entry.entry_id
```
Please use entry.runtimedata

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,69 @@
+    hass.data[DOMAIN][entry.entry_id] = coordinator
+    hass.data[DOMAIN]["entry_id"] = entry.entry_id
+
+    await coordinator.async_config_entry_first_refresh()
```
Please refresh the coordinator before assigning to entry.runtimedata

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,69 @@
+    return unload_ok
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: MadVRConfigEntry) -> None:
+    """Reload a config entry."""
+    await async_unload_entry(hass, entry)
+    await async_setup_entry(hass, entry)
```
uneeeded

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,153 @@
+            keep_power_on = user_input["keep_power_on"]
+            try:
+                await self._test_connection(host, port, mac, keep_power_on)
+                return self.async_create_entry(
+                    title=user_input["name"], data=user_input
+                )
```
Only keep things int he try block that can raise

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,153 @@
+            except CannotConnect:
+                _LOGGER.error("CannotConnect error caught")
+                errors["base"] = "cannot_connect"
+                self.context["user_input"] = user_input
```
why this

iloveicedgreentea replied:
doesnt this display the error in the UI?

joostlek replied:
no the self.context

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,153 @@
+                _LOGGER.error("CannotConnect error caught")
+                errors["base"] = "cannot_connect"
+                self.context["user_input"] = user_input
+                # allow user to skip connection test
+                return await self.async_step_confirm()
```
why? The config flow is the userfriendliest way to tell the user they fucked up and thus something is wrong

iloveicedgreentea replied:
This specific part was added because it may be the case the device is not connected yet or set up but the integrator knows the connection info already, or maybe the device isn't plugged in, etc. I want to give the user freedom to move on and add anyway as they can modify the options after adding it

joostlek replied:
We should confirm we can connect with the device before the user is able to setup. This way we can ensure everything works like expected.

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,153 @@
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required("name"): str,
+                    vol.Required("host"): str,
+                    vol.Required("mac"): str,
+                    vol.Optional("port", default=44077): int,
+                    vol.Optional("keep_power_on", default=False): bool,
```
Why does `__init__.py` take into account that these values are optional as they always get a value from the looks of it

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,153 @@
+            data_schema=vol.Schema({vol.Required("confirm", default=False): bool}),
+        )
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry: ConfigEntry) -> OptionsFlow:
+        """Get the options flow for this handler."""
+        return MadVROptionsFlowHandler(config_entry)
```
Follow up

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,80 @@
+
+
+class MadVRCoordinator(DataUpdateCoordinator):
+    """My custom coordinator for push-based API."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
```
```suggestion
    """My custom coordinator for push-based API."""
    
    config_entry: ConfigEntry

    def __init__(
        self,
        hass: HomeAssistant,
```

joostlek replied:
You can then access `self.config_entry` afterwards

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,80 @@
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="My API Coordinator",
```
Stale

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,80 @@
+        self.my_api.set_update_callback(self.handle_push_data)
+        _LOGGER.debug("MadVRCoordinator initialized")
+
+    async def _async_update_data(self):
+        """No-op method for initial setup."""
+        return self.previous_data
```
euh?

iloveicedgreentea replied:
I had a HA error unless I added it but could have been caused by something else. I will try to remove

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,80 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class MadVRCoordinator(DataUpdateCoordinator):
```
Please type the coordinator to the type of `data`
```suggestion
class MadVRCoordinator(DataUpdateCoordinator[..]):
```

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,80 @@
+                    "new_state": data.get("is_on"),
+                    "old_state": self.previous_data.get("is_on"),
+                }
+                self.hass.bus.async_fire(f"{DOMAIN}_event", event_data)
+
+            # Compare HDR flag
+            if data.get("hdr_flag") != self.previous_data.get("hdr_flag"):
+                event_data = {
+                    "device_id": self.entry_id,
+                    "type": "hdr_flag_changed",
+                    "new_state": data.get("hdr_flag"),
+                    "old_state": self.previous_data.get("hdr_flag"),
+                }
+                self.hass.bus.async_fire(f"{DOMAIN}_event", event_data)
+
+            # Compare aspect ratio
+            if data.get("aspect_dec") != self.previous_data.get("aspect_dec"):
+                event_data = {
+                    "device_id": self.entry_id,
+                    "type": "aspect_dec_changed",
+                    "new_state": data.get("aspect_dec"),
+                    "old_state": self.previous_data.get("aspect_dec"),
+                }
+                self.hass.bus.async_fire(f"{DOMAIN}_event", event_data)
```
Let's not use the event bus here, what are we doing here

iloveicedgreentea replied:
The point of this is to send events when these values change. This is what the docs say to do. Is there a better way to handle this?

joostlek replied:
I am asking a more global question, what does this do and how does the integration communicate with the device and vice versa

iloveicedgreentea replied:
It opens a persistent connection to the device and the device pushes data to the integration. The integration processes that data in realtime and updates its internal state. So the integration reflects the real time state of the device

This code would fire events on the bus so users can create automations based on specific states of the device in realtime if they would like an event based system. I supposed they can also use sensor data (had to remove it because the bot wont let me add more than 1 platform but will add later) so I guess I will remove this for simplicity

joostlek replied:
Yes but we want to avoid using the bus as much as possible if we can create entities for them.

What are the types of events we receive? What's the state?

iloveicedgreentea replied:
I removed this code

joostlek commented on homeassistant/components/madvr/wakeonlan.py:
We should not add this. the user can set it up themselves

iloveicedgreentea replied:
This library is used to wake the device which otherwise is impossible to wake up. Is there a built in for this? Its unreasonable for a user to set up wakeonlan first then install this integration. Alternatively, I can bake this into my library and call it from the integration

joostlek replied:
The user should use wake on lan for this. The library should not do this

iloveicedgreentea committed:
fix: use conf, refresh coordinator, other comments

iloveicedgreentea committed:
fix: remove event data

iloveicedgreentea committed:
fix: fix tests passing, remove wake on lan

iloveicedgreentea committed:
fix: dont allow to proceed unless connection works

iloveicedgreentea committed:
chore: update dep

iloveicedgreentea commented:
Wake on lan removed, events removed, other comments addressed. I put this code into a custom component and test on my prod instance, all features work as expected too. (power on/off, attributes, sending commands, etc)

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,67 @@
+from .coordinator import MadVRCoordinator
+from .utils import cancel_tasks
+
+# For your initial PR, limit it to 1 platform.
```
Can eb removed

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,67 @@
+    madVRClient = Madvr(
+        host=entry.data[CONF_HOST],
+        logger=_LOGGER,
+        port=entry.data.get(CONF_PORT, 44077),
```
Port is not optional

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+                }
+            ),
+            errors=errors,
+            description_placeholders={
+                "error_message": "Failed to connect. Please check your settings and ensure the device is on."
+                if errors
+                else ""
+            },
```
Why is this?

iloveicedgreentea replied:
removed

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+            host = user_input[CONF_HOST]
+            port = user_input.get(CONF_PORT, 44077)
+            mac = user_input[CONF_MAC]
+            keep_power_on = user_input["keep_power_on"]
```
What does this setting mean

iloveicedgreentea replied:
During adding the device, it has to be on to be connected. If the user wishes, the integration can turn it off after or keep it on

iloveicedgreentea replied:
moved to a const also

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+class MadVRConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for the integration."""
+
+    async def async_step_user(self, user_input=None) -> ConfigFlowResult:
```
Please type appropraitly

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class MadVRConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class MadVRConfigFlow(ConfigFlow, domain=DOMAIN):
```

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+        # Whether it's the first attempt or a retry, show the form
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_NAME,
+                        default=user_input.get(CONF_NAME, "") if user_input else "",
+                    ): str,
+                    vol.Required(
+                        CONF_HOST,
+                        default=user_input.get(CONF_HOST, "") if user_input else "",
+                    ): str,
+                    vol.Required(
+                        CONF_MAC,
+                        default=user_input.get(CONF_MAC, "") if user_input else "",
+                    ): str,
+                    vol.Required(
+                        CONF_PORT,
+                        default=user_input.get(CONF_PORT, 44077)
+                        if user_input
+                        else 44077,
+                    ): int,
+                    vol.Required(
+                        "keep_power_on",
+                        default=user_input.get("keep_power_on", False)
+                        if user_input
+                        else False,
+                    ): bool,
+                }
```
Please use `self.add_suggested_values` to add suggested values. Please also move the data schema out of the ConfigFlow. Check `pyload` for an example

iloveicedgreentea replied:
thank you for the example

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,161 @@
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_NAME,
+                        default=user_input.get(CONF_NAME, "") if user_input else "",
+                    ): str,
```
Name is only allowed to be user configurable when its a helper

joostlek reviewed (changes requested):
Is it possible to maybe start with a platform that holds a state? Currently its very unclear what data you have, what you work with and that will affect how it is going to look in the end.

Also, not all comments are fixed from the previous round.

jbouwh commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,176 @@
+        task_hb = self.hass.loop.create_task(self.madvr_client.send_heartbeat())
+        self.tasks.append(task_hb)
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when removed."""
+        _LOGGER.debug("Removing from hass")
+        self.madvr_client.stop()
+        await self.madvr_client.close_connection()
+        await cancel_tasks(self.madvr_client)
+        for task in self.tasks:
+            if not task.done():
+                task.cancel()
```
This is also for the coordinator entity

jbouwh commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,176 @@
+                    self.command_queue.task_done()
+
+            if self.stop_processing_commands.is_set():
+                await self.clear_queue()
```
Why is this awaited?

iloveicedgreentea replied:
because the method is async

jbouwh replied:
Nothing is awaited in the coro

jbouwh commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,176 @@
+                    await self.madvr_client.send_command(command)
+                except ConnectionResetError:
+                    _LOGGER.warning("Envy was turned off manually")
+                    await self.async_turn_off()
```
If the state is `OFF` please update the state and write the entities state. Further it seems not logical to turn off a device if it cannot be connected at this stage.

iloveicedgreentea replied:
updated this and will split out the functionality to disconnect and reset from "power off". The intention here is to mark the device as off

jbouwh commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,176 @@
+        self.stop_processing_commands.set()
+        await self.clear_queue()
+        await self.madvr_client.power_off()
+        self._attr_is_on = False
```
If you change the entities state, you should write it as well.

jbouwh commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,176 @@
+            "new_state": "on",
+            "old_state": "off",
+        }
+        self.hass.bus.async_fire(f"{DOMAIN}_event", event_data)
```
What is this used for?

iloveicedgreentea replied:
that should have been removed, will remove this

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,176 @@
+        return self.madvr_client.is_on
+
+    @property
+    def should_poll(self) -> bool:
+        """Poll."""
+        return False
+
```
It's using a coordinator so this is the default

iloveicedgreentea committed:
fix: update config flow, add constants

iloveicedgreentea commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,176 @@
+        self.stop_processing_commands = asyncio.Event()
+        self.madvr_client.set_update_callback(coordinator.handle_push_data)
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        # open connection if not already connected
+        await super().async_added_to_hass()
+
+        # handle queue
+        task_queue = self.hass.loop.create_task(self.handle_queue())
+        self.tasks.append(task_queue)
+
+        task_notif = self.hass.loop.create_task(self.madvr_client.read_notifications())
+        self.tasks.append(task_notif)
+
+        task_hb = self.hass.loop.create_task(self.madvr_client.send_heartbeat())
+        self.tasks.append(task_hb)
```
These tasks must only run when the entity is added though. Is that still the case if its moved to coordinator?

jbouwh replied:
The entity connects to the coordinator when it loads en fetches it's data to supply all entities, In your case there is only one entity yet.

The entity code should not connect, the coordinator should do this.
https://developers.home-assistant.io/docs/integration_fetching_data/

iloveicedgreentea commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,176 @@
+        self._attr_name = coordinator.name
+        self._attr_unique_id = f"{entry_id}_remote"
+        self.entry_id = hass.data[DOMAIN]["entry_id"]
+        self._attr_should_poll = False
```
I don't fully understand what you mean can you please elaborate?

> Could move this to a coordinator entity and inherit that one

jbouwh replied:
This comment is not relevant for this PR.

iloveicedgreentea committed:
fix: write state, use runtime data instead

iloveicedgreentea committed:
fix: remove await

iloveicedgreentea commented:
Changes:

* Use runtime data
* Split out schema from config flow
* Add types to missing params
* Remove lingering event bus stuff
* Remove unneeded methods
* Write state after changing it
* Fix misc comments

>Is it possible to maybe start with a platform that holds a state? Currently its very unclear what data you have, what you work with and that will affect how it is going to look in the end.

Not sure what you mean. The library handles receiving push data from the device. This integration reads self.madvr_client.msg_dict

Power state is read directly from the device via the library in self.madvr_client.is_on

Otherwise, it sends commands to the device. This works perfectly fine I have been testing this integration for 3 weeks now and the library itself has been used by many people for about a year now. I used most of this code in a custom component for over a year too.

joostlek reviewed (commented):
I think its good to take a step back at this point. How does the data work in the integration? What does the state of the data look like?

Some things don't make sense and I think we better take a step back and increase the shared knowledge before we send you in a direction that does not work. Because you are creating a MadVR client in the `__init__.py`, and then you are actually closing the connection the moment the remote is removed from hass (in `async_will_remove_from_hass`) even though this also happens when someone disabled that entity. So this would lead to unwanted side effects.

I also think that we should explore putting the whole queue mechanism into the library, as I think that is a device/connection specific detail.

iloveicedgreentea commented:
>I think its good to take a step back at this point. How does the data work in the integration? What does the state of the data look like?

I still really don't understand this comment. The integration literally reads a dict in the library which is written by the library when the device pushes data through the connection, like a websocket. This is how the device is designed. 

>Some things don't make sense and I think we better take a step back and increase the shared knowledge before we send you in a direction that does not work. Because you are creating a MadVR client in the __init__.py, and then you are actually closing the connection the moment the remote is removed from hass (in async_will_remove_from_hass) even though this also happens when someone disabled that entity. So this would lead to unwanted side effects.

I think what doesn't make sense is the HA dev docs. It is frustratingly hard to understand "the right way" to do something. All the guidance given in this PR is not found in the docs and I have just been following the docs to create this integration. Those functions don't have explanations like that

So are you saying to handle connection closing in async_unload_entry? You need to understand I don't have the institutional knowledge you have. I only have access to the docs. And the docs say that async_will_remove_from_hass is used to "disconnect from the server" The docs also say async_unload_entry is used to "close all connections" so which one do I use? The functions seem to do the exact same thing.

>I also think that we should explore putting the whole queue mechanism into the library, as I think that is a device/connection specific detail.

This is fair I will move this to the library

joostlek commented:
> The integration literally reads a dict in the library which is written by the library when the device pushes data through the connection, like a websocket. This is how the device is designed

I did not know this. I could only assume and since in this case I was confused with all the things happening with tasks and stuff I just asked the quest to be sure :)

>I think what doesn't make sense is the HA dev docs. It is frustratingly hard to understand "the right way" to do something. All the guidance given in this PR is not found in the docs and I have just been following the docs to create this integration. Those functions don't have explanations like that

1. The dev docs are lacking and I acknowledge that. It's something I want to work on in the near future since I also only gained this knowledge in the last year.
2. There is no right way for a lot of things as its integration dependent. I haven't seen an integration that is exactly like another one. But there are patterns to be seen.

> And the docs say that async_will_remove_from_hass is used to "disconnect from the server" The docs also say async_unload_entry is used to "close all connections" so which one do I use? The functions seem to do the exact same thing.

In the past, a lot of integrations just had 1 entity which contained a lot of data. In the recent times we've seen a move towards more entities instead of more data in 1 entity, as the first one is more user friendly. But the docs haven't really grown with that. So yea, in the past it was more common to disconnect at that place, since with YAML there was no way of unloading. So yes, use `async_unload_entry` as that is integration scale instead of entity scale.

And please don't forget, I am not here to call you out on stuff you did wrong, I am trying to point you at the relevant things that could increase the code quality and enhancing best practices. I am currently looking into how we can word best practices and present them as a guide to enhance the steps to create a new integration.

iloveicedgreentea committed:
fix: move unloading and stuff to coordinator/init

iloveicedgreentea commented:
>And please don't forget, I am not here to call you out on stuff you did wrong, I am trying to point you at the relevant things that could increase the code quality and enhancing best practices. I am currently looking into how we can word best practices and present them as a guide to enhance the steps to create a new integration.

Thats fine actually I don't mind being corrected because I want to do it the right way. I do think that, with complex systems, there is a "right way" and I would love for HA to be more opinionated with that. HA should prescribe ways to do this with patterns rather than documenting functions. For example, in my case I am using a push method, so there should be one way to do this. Having documentation that points to existing code bases which are exemplars of this would be a good start. This is the feedback I see anyway, that people need to dive into core and just blindly copy existing integrations.

e.g in the getting started docs, have a section with common patterns:

Is your library pushing data to the integration? Here is how you set that up and here is an exemplar integration

Anyway, I pushed a change which moves more things into the coordinator and redid the unload stuff. Hopefully that is right. I am working on moving the queue stuff into the library

iloveicedgreentea commented:
>I did not know this. I could only assume and since in this case I was confused with all the things happening with tasks and stuff I just asked the quest to be sure :)

I mentioned it before but I can add this to comments in the code if that will help. The tasks are used to process things in real time

1. Handle queue (moved to library) this processes commands in real time asynchronously to send them to device
2. Heartbeat. The device requires a heartbeat within 30 seconds, but I can move this to library also
3. read_notifications. This handles processing push notifications in real time.
Now that I think about it, I think all three of these can be moved to the library but their lifecycle can be managed by the integration so they start and stop correctly. I will update this.

>And please don't forget, I am not here to call you out on stuff you did wrong, I am trying to point you at the relevant things that could increase the code quality and enhancing best practices. I am currently looking into how we can word best practices and present them as a guide to enhance the steps to create a new integration.

Thats fine actually I don't mind being corrected because I want to do it the right way. I do think that, with complex systems, there is a "right way" and I would love for HA to be more opinionated with that. HA should prescribe ways to do this with patterns rather than documenting functions. For example, in my case I am using a push method, so there should be one way to do this. Having documentation that points to existing code bases which are exemplars of this would be a good start. This is the feedback I see anyway, that people need to dive into core and just blindly copy existing integrations.

e.g in the getting started docs, have a section with common patterns:

Is your library pushing data to the integration? Here is how you set that up and here is an exemplar integration

Anyway, I pushed a change which moves more things into the coordinator and redid the unload stuff. Hopefully that is right. I am working on moving the queue stuff into the library

joostlek commented:
> I think all three of these can be moved to the library but their lifecycle can be managed by the integration so they start and stop correctly. I will update this.

Awesome!

I mean if you have a lot of passion for docs, feel free to contact me on discord so I have someone new ish who I can throw it to, to check if its understandable :P

Let me know when stuff is done, then we can take a look at it again :)

iloveicedgreentea committed:
fix: pass in config entry with correct type

iloveicedgreentea committed:
fix: move queue and tasks to library

iloveicedgreentea committed:
fix: config flow error flow, tests, name, and update lib

iloveicedgreentea committed:
fix: update lib, leave connection open on setup

iloveicedgreentea committed:
fix: update lib

iloveicedgreentea commented:
> > I think all three of these can be moved to the library but their lifecycle can be managed by the integration so they start and stop correctly. I will update this.
> 
> Awesome!
> 
> I mean if you have a lot of passion for docs, feel free to contact me on discord so I have someone new ish who I can throw it to, to check if its understandable :P
> 
> Let me know when stuff is done, then we can take a look at it again :)

Yes I would definitely like to help out. I would like to be involved in HA development more so this is a good first step. I can ping you on discord

I pushed some updates

1. Moved queues and tasks into library
2. corrected/updated type hints
3. removed name from config
4. fixed test failures
5. updated library to be more efficient and process updates better
6. added mypy to library to catch any silly stuff

I tested this locally and on my prod instance, everything works as expected (attributes, push updates, tasks, reload & disabling integration, etc)

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,67 @@
+        madVRClient,
+        name=name,
+    )
+    hass.data.setdefault(DOMAIN, {})
```
Unneeded

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,67 @@
+        name=name,
+    )
+    hass.data.setdefault(DOMAIN, {})
+    await coordinator.async_refresh()
```
```suggestion
    await coordinator.async_config_entry_first_refresh()
```

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,67 @@
+    hass.data.setdefault(DOMAIN, {})
+    await coordinator.async_refresh()
+    entry.runtime_data = coordinator
+    hass.data[DOMAIN]["entry_id"] = entry.entry_id
+
+    await coordinator.async_config_entry_first_refresh()
```
```suggestion
```

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,67 @@
+
+    await coordinator.async_config_entry_first_refresh()
+
+    # Forward the entry setup to the platforms
```
```suggestion
```

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,174 @@
+            try:
+                await self._test_connection(host, port, mac, keep_power_on)
+                return self.async_create_entry(
+                    title=user_input.pop(CONF_NAME, DEFAULT_NAME),
```
name is not present in here

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,67 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: MadVRConfigEntry) -> bool:
+    """Set up the integration from a config entry."""
+    name = entry.data.get(CONF_NAME, DEFAULT_NAME)
```
theres no name in there

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,56 @@
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: MadVRConfigEntry,
```
```suggestion
```

after the `super().__init__(` in this file, you can just freely access `self.config_entry`

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,56 @@
+type MadVRConfigEntry = ConfigEntry[MadVRCoordinator]
+
+
+class MadVRCoordinator(DataUpdateCoordinator[dict]):
```
Can we improve the dict typing?

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,56 @@
+        hass: HomeAssistant,
+        config_entry: MadVRConfigEntry,
+        client: Madvr,
+        name: str,
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Madvr Coordinator",
+        )
```
not sure why we pass name in here now

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,56 @@
+        """No-op method for initial setup."""
+        return
+
+    def handle_push_data(self, data: dict):
```
missing return type annotation

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,56 @@
+        _LOGGER.debug("Received push data: %s", data)
+        self.async_set_updated_data(data)
+
+    async def async_handle_unload(self):
```
missing return type annotation

joostlek replied:
I think you could even move this content to either `async_unload_entry` (because I think it makes sense since `__init__.py` creates the client, so it also has to take care of unloading the client). You could also make use of callbacks, an example:
https://github.com/home-assistant/core/blob/540da3cac6fef9d2fe691ead9718142129ed5832/homeassistant/components/nanoleaf/__init__.py#L80-L87

joostlek commented on homeassistant/components/madvr/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "name": "MadVR Envy",
+  "codeowners": ["@iloveicedgreentea"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://www.home-assistant.io/integrations/madvr",
+  "homekit": {},
```
Please remove empty fields

joostlek commented on homeassistant/components/madvr/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "homekit": {},
+  "integration_type": "device",
+  "iot_class": "local_push",
+  "quality_scale": "silver",
```
not yet :)

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,106 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the MadVR remote."""
+    coordinator: MadVRCoordinator = entry.runtime_data
+    if not isinstance(coordinator, MadVRCoordinator):
+        raise TypeError("entry.runtime_data is not an instance of MadVRCoordinator")
```
```suggestion
    hass: HomeAssistant,
    entry: MadVRConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up the MadVR remote."""
    coordinator = entry.runtime_data
```

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,106 @@
+        self.coordinator = coordinator
+        self._attr_name = coordinator.name
+        self._attr_unique_id = f"{entry_id}_remote"
+        self.entry_id = hass.data[DOMAIN]["entry_id"]
```
```suggestion
```

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,174 @@
+                "Testing connection to MadVR at %s:%s with mac %s", host, port, mac
+            )
+            # turn on the device
+            await madvr_client.power_on()
```
Does this now still use Wake on Lan? We should not do that, its not a strange requirement to have people turn on their device before being able to add it

iloveicedgreentea replied:
wol removed

joostlek replied:
How doe turning on now work?

iloveicedgreentea replied:
without wake on lan, it is not possible. It must have wake on lan implemented to turn it on

joostlek replied:
So you're saying you still use wake on Lan?

iloveicedgreentea replied:
no if a mac is not provided the library will not do anything when the power command is called. It is impossible

joostlek replied:
But then why do we implement something that does not work?

iloveicedgreentea replied:
it doesnt work because you wont let me include wake on lan inside the library :)

it will work once device trigger is implemented, but Im trying to tell you it is _impossible_ to action this device without wake on lan and not being able to call power on is a confusing user experience

joostlek replied:
But then we should not implement `async_turn_on` if we can't turn it on via the library and not just implement a library call that doesn't work

iloveicedgreentea replied:
okay I will remove for now then

iloveicedgreentea replied:
actually a counter proposition

The original complaint was entering a mac is bad. So how about during initial config, asking the user to turn it on and connecting, then I can make a call to the device to get the mac and store it, then use wake on lan (inside the library) to control it

This seems like best of both worlds and letting the device handle the power lifecycle and also not asking user for mac input

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,174 @@
+        # wait for it to be available (envy takes about 15s to boot)
+        await asyncio.sleep(15)
+        # try to connect
+        try:
+            await asyncio.wait_for(madvr_client.open_connection(), timeout=15)
+        # connection can raise HeartBeatError if the device is not available or connection does not work
+        except (TimeoutError, aiohttp.ClientError, OSError, HeartBeatError) as err:
+            _LOGGER.error("Error connecting to MadVR: %s", err)
+            raise CannotConnect from err
```
Would this also be something for the library?

joostlek replied:
Unanswered

iloveicedgreentea replied:
what part specifically? I do not understand the question

joostlek replied:
The whole, waiting for the device to respond within 15 seconds otherwise raise error

iloveicedgreentea replied:
refactored this

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,174 @@
+        """Get the options flow for this handler."""
+        return MadVROptionsFlowHandler(config_entry)
+
+    async def async_step_import(
+        self, import_config: dict[str, Any]
+    ) -> ConfigFlowResult:
+        """Import a config entry from configuration.yaml."""
+        config = {
+            CONF_NAME: import_config.get(CONF_NAME, DEFAULT_NAME),
+            CONF_HOST: import_config.get(CONF_HOST, ""),
+            CONF_PORT: import_config.get(CONF_PORT, DEFAULT_PORT),
+            CONF_MAC: import_config.get(CONF_MAC, ""),
+            CONF_POWER_ON: False,
+        }
+
+        result = await self.async_step_user(config)
+
+        if errors := result.get("errors"):
+            return self.async_abort(reason=errors["base"])
+        return result
```
Why do we have an import step?

joostlek commented on homeassistant/components/madvr/config_flow.py:
Is there something unique we can get from the device? I see the user has to fill in a mac address, but I am assuming this is required for Wake on LAN which we should not do. having to enter a mac address is a terrible user experience. (because do you have to include `:` , does it have to be uppercase? nobody knows).

So can we get something unique from the device instead?

iloveicedgreentea replied:
mac removed

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,106 @@
+    )
+
+
+class MadvrRemote(CoordinatorEntity, RemoteEntity):
+    """Remote entity for the MadVR integration."""
+
+    coordinator: MadVRCoordinator
```
```suggestion
class MadvrRemote(CoordinatorEntity[MadVRCoordinator], RemoteEntity):
    """Remote entity for the MadVR integration."""
```

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,106 @@
+        self,
+        hass: HomeAssistant,
+        coordinator: MadVRCoordinator,
+        madvr_client: Madvr,
```
Why are we passing this in? we can easily access it from the coordinator

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,106 @@
+        self.entry_id = hass.data[DOMAIN]["entry_id"]
+        self._attr_should_poll = False
+        self.connection_event = self.madvr_client.connection_event
+        self.madvr_client.set_update_callback(coordinator.handle_push_data)
```
Why do we register this callback twice?

iloveicedgreentea replied:
you are right, removed

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,106 @@
+        # inform the client to start tasks
+        _LOGGER.debug("Adding to hass")
+        _LOGGER.debug("Using loop: %s", self.madvr_client.loop)
+        await self.madvr_client.async_add_tasks()
```
So what does this do exactly?

iloveicedgreentea replied:
this informs the library it should start the background tasks so the integration manages the lifecycle. They will also be cancelled on unload

1. Ping device to determine power state
2. Heartbeat for connection to keepalive
3. Handle commands queue
4. process push notifications from device async

joostlek replied:
Let's do that not in the entity

iloveicedgreentea replied:
in the coordinator I assume? What are the lifecycle functions for the coordinator I should use?

joostlek replied:
There are none yet

iloveicedgreentea replied:
okay so how would I move this outside the entity?

joostlek replied:
Create your own function in the coordinator and call that after creating the coordinator

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,106 @@
+        """Run when removed."""
+        _LOGGER.debug("Removing from hass")
+        # inform the client to cancel all tasks
+        await self.coordinator.async_handle_unload()
```
So this destroys the connection the device, but we already do that on integration level so I don't think we should do this in entity level as well

iloveicedgreentea replied:
> but we already do that on integration level

Where?

joostlek replied:
Async unload entry

iloveicedgreentea replied:
yes you are right, removed from entity

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,106 @@
+        """Return true if the device is on."""
+        return self.madvr_client.is_on
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any]:
+        """Return extra state attributes."""
+        # Useful for making sensors
+        return self.madvr_client.msg_dict
```
If they are useful for making sensors, let's make them proper sensors instead. We want to add as little unneeded extra state attributes as possible

iloveicedgreentea replied:
okay I can remove this. The original PR had sensors and binary sensors but the bot made me remove them

joostlek commented on homeassistant/components/madvr/strings.json:
```diff
@@ -0,0 +1,36 @@
+        "title": "Setup MadVR Envy",
+        "description": "Your device will be turned on so we can test the connection. ",
+        "data": {
+          "host": "Host or IP Address",
+          "mac": "MAC Address",
+          "port": "Port",
```
Please use references, see other integrations on this

iloveicedgreentea committed:
fix: address comments, remove wol from lib

iloveicedgreentea committed:
fix: remove unneeded options

iloveicedgreentea commented:
1. Removed extra attributes (will make PR to add sensors, binary sensors after this)
2. used common string references
3. removed duplicate unload logic 
4. moved task and related logic to coordinator
5. addressed other misc comments
6. removed mac and WOL (intention is to set up device trigger in new PR)
7. changed library logic to use standby instead of power off (removes WOL dependency until its implemented)
8. simplified config flow

iloveicedgreentea committed:
fix: remove fields

joostlek commented:
Please go through the old comments, there are some thing that I commented on and haven't been resolved or answered

iloveicedgreentea committed:
fix: simplify code, address comments

iloveicedgreentea committed:
fix: move error to lib

iloveicedgreentea committed:
fix: fix test

iloveicedgreentea committed:
fix: stronger types

iloveicedgreentea commented:
> Please go through the old comments, there are some thing that I commented on and haven't been resolved or answered

sorry they didnt appear in the conversation view so I missed it

Addressed the comments that I saw

joostlek commented:
There are more :)
config_flow.py

iloveicedgreentea committed:
fix: update lib

iloveicedgreentea commented:
1. Added better typing to stuff
2. Change wol behavior in the library to not require mac
3. Moved some components into library
4. Retested behavior (config flow, disable, reload, power stuff, etc)

joostlek commented:
The options flow needs to go in the initial PR

iloveicedgreentea committed:
fix: missing text from options flow

iloveicedgreentea commented:
> The options flow needs to go in the initial PR

Not sure what this means. Options flow is defined here https://github.com/home-assistant/core/pull/120382/files#diff-cc12f13f6b47fdf0c245334572e9c5ca50f877b5d6ae97679340ae55df3672a7R121

joostlek commented:
Sorry, typo, options flow needs to go in a follow up PR

iloveicedgreentea committed:
chore: remove options flow

iloveicedgreentea commented:
okay options flow is removed

iloveicedgreentea committed:
chore: remove import

iloveicedgreentea committed:
chore: update comments

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    )
+    coordinator = MadVRCoordinator(hass, madVRClient)
+
+    await coordinator.async_config_entry_first_refresh()
```
Right, so the integration is pull based so this actually does nothing. Is there a way we can still confirm the device is present at that place and that we are able to succeed with setting up?

iloveicedgreentea replied:
>Is there a way we can still confirm the device is present at that place and that we are able to succeed with setting up?

only way to confirm the device is to connect to it, and it must be on to do so

joostlek replied:
Where do we connect with it exactly?

iloveicedgreentea replied:
async_add_tasks starts the library background workers

once the library detects the device is pingable, it automatically initiates a connection (if its not connected already)

It also opens a connection in config flow after testing it and maintains that

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    entry.runtime_data = coordinator
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    entry.add_update_listener(async_reload_entry)
```
Not needed anymore

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        coordinator: MadVRCoordinator = entry.runtime_data
+        if coordinator:
```
coordinator is not nullable

joostlek commented on homeassistant/components/madvr/__init__.py:
```diff
@@ -0,0 +1,68 @@
+    return True
+
+
+async def async_handle_unload(coordinator: MadVRCoordinator) -> None:
```
Let's also listen to HOME_ASSISTANT_STOP on the event bus and execute this function

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+            port = user_input[CONF_PORT]
+            try:
+                await self._test_connection(host, port)
+                return self.async_create_entry(
+                    title=DEFAULT_NAME,
+                    data=user_input,
+                )
```
Only have things in the try block that can raise

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+            errors=errors,
+        )
+
+    async def _test_connection(self, host: str, port: int):
```
Please fully type

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,50 @@
+        """No-op method for initial setup."""
+        return
+
+    def handle_push_data(self, data: dict) -> None:
```
Please fully type

joostlek commented on homeassistant/components/madvr/manifest.json:
```diff
@@ -0,0 +1,11 @@
+  "name": "MadVR Envy",
+  "codeowners": ["@iloveicedgreentea"],
+  "config_flow": true,
+  "dependencies": [],
```
Remove empty fields

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,87 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+type MadVRConfigEntry = ConfigEntry[MadVRCoordinator]
```
Import this from `.`

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,50 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+type MadVRConfigEntry = ConfigEntry[MadVRCoordinator]
```
Idem
You might need a `if TYPE_CHECKING:` block

iloveicedgreentea replied:
I had this before but then you said to remove it. If I use if TYPE_CHECKING here, I have to use it below which you wanted me to remove

joostlek replied:
ooooooh

```py
if TYPE_CHECKING:
    from . import MadVRConfigEntry
```
You also probably have to add `from __future__ import annotations`

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,87 @@
+) -> None:
+    """Set up the MadVR remote."""
+    coordinator = entry.runtime_data
+    if not isinstance(coordinator, MadVRCoordinator):
+        raise TypeError("entry.runtime_data is not an instance of MadVRCoordinator")
```
Why?

iloveicedgreentea replied:
had mypy issues since resolved, will remove

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,87 @@
+    ) -> None:
+        """Initialize the remote entity."""
+        super().__init__(coordinator)
+        self.coordinator = coordinator
```
Unneeded

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,87 @@
+        super().__init__(coordinator)
+        self.coordinator = coordinator
+        self.madvr_client = self.coordinator.client
+        self._attr_name = self.coordinator.name
```
Please use `device_info`

iloveicedgreentea replied:
do you have an example of what you mean

joostlek replied:
About 700 integrations, look for the usage of DeviceInfo and look at the dev docs

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,87 @@
+        self.madvr_client = self.coordinator.client
+        self._attr_name = self.coordinator.name
+        self._attr_unique_id = f"{entry_id}_remote"
+        self.entry_id = entry_id
```
why?

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,87 @@
+        self._attr_name = self.coordinator.name
+        self._attr_unique_id = f"{entry_id}_remote"
+        self.entry_id = entry_id
+        self._attr_should_poll = False
```
Can be set outside of the constructor

iloveicedgreentea replied:
where?

joostlek replied:
Outside of the constructor as a class variable

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,87 @@
+        self._attr_should_poll = False
+        self.connection_event = self.madvr_client.connection_event
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        await super().async_added_to_hass()
+        _LOGGER.debug("Adding to hass")
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when removed."""
+        _LOGGER.debug("Removing from hass")
```
Can both be removed

joostlek commented on homeassistant/components/madvr/strings.json:
```diff
@@ -0,0 +1,33 @@
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]"
+    }
+  },
+  "options": {
```
Unneeded

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+)
+
+
+class MadVRConfigFlow(ConfigFlow, domain=DOMAIN):
```
Again the question, is there something unique we can get from the device? serial number, macaddress etc?

iloveicedgreentea replied:
once the connection is established it is possible to ask it for a mac address but nothing before a connection is established. What is the purpose of unique info here?

joostlek replied:
Making sure someone can't setup the same config flow twice

iloveicedgreentea replied:
before it was asking for a name and mac, but I was told to remove that. So I really don't know what else to do here.

joostlek replied:
Oh I misread that, I read that we could retrieve the Mac.

Can we acquire something from the device that is unique after connecting?

iloveicedgreentea replied:
nothing besides a mac there is no command to get any unique info

Could ask the user to input their serial number but at the same time, it doesn't really matter if you set it up twice because the connections will be completely independent, I tested this with multiple instances of the integration

joostlek replied:
Yea but it's a user experience thing. In that case we should abort if we see the same host.

iloveicedgreentea replied:
okay I added 

```
await self.async_set_unique_id(host)
self._abort_if_unique_id_configured()
```

iloveicedgreentea committed:
fix: get mac from device, persist

iloveicedgreentea committed:
fix: add mac stuff to test

iloveicedgreentea committed:
fix: startup import errors

iloveicedgreentea committed:
chore: stale comment

iloveicedgreentea committed:
fix: get mac from persisted config

iloveicedgreentea commented:
1. Balancing device limitations, I added library functionality to grab the mac from the device. The user must have the device on during config, then the mac will be persisted in config entry to be passed to the library. This makes power on work without needing to enter in mac
2. abort if duplicate host found
3. remove unnecessary functions

iloveicedgreentea committed:
chore: update lib

iloveicedgreentea commented:
Anything else needed?

iloveicedgreentea committed:
fix: persist mac in a better way

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+        await asyncio.sleep(2)
+        # get mac from device
+        mac_address = madvr_client.mac_address
+        if not mac_address:
+            # its not critical because it should get picked up by the client eventually but not ideal
```
I mean, having a Mac address is nice because you can use it for the unique ID of the config flow, making sure people can't set up the integration twice. You can also use the mac address in the entity unique ids. This means if someone has the integration set up and has automations attached to it, and they have to set up the integration again for some reason, they don't have to reconfigure every automation again so I highly recommend getting the Mac address from the device

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+class MadVRCoordinator(DataUpdateCoordinator[dict[str, Any]]):
+    """Madvr coordinator for Envy (push-based API)."""
+
+    if TYPE_CHECKING:
```
We usually don't do the if type checking here, not sure what the impact is but if we can remove it, that would be nice

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+            identifiers={(DOMAIN, self.config_entry.entry_id)},
+            default_manufacturer="madVR",
+            name="Envy",
+            connections={
+                (CONF_MAC, self.config_entry.data.get(CONF_MAC, "")),
+                (CONF_HOST, self.config_entry.data[CONF_HOST]),
+                (CONF_PORT, self.config_entry.data[CONF_PORT]),
+            },
```
That's not how the connections work

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+        _LOGGER.debug("MadVRCoordinator initialized with mac: %s", self.mac)
+
+    @property
+    def device_info(self) -> DeviceInfo:
```
I think we can move this into a shared base entity in the future, for now you can also just keep it in the remote

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,74 @@
+class MadvrRemote(CoordinatorEntity[MadVRCoordinator], RemoteEntity):
+    """Remote entity for the MadVR integration."""
+
+    _attr_should_poll = False
```
Is default with coordinator entities, so can be removed

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,74 @@
+        """Initialize the remote entity."""
+        super().__init__(coordinator)
+        self.madvr_client = coordinator.client
+        self._attr_name = coordinator.device_info.get("name")
```
Integrations should have `_attr_has_entity_name = True`, which prepends the entity name with the device name. To make this entity be the device name, add `_attr_name = None`

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,74 @@
+        super().__init__(coordinator)
+        self.madvr_client = coordinator.client
+        self._attr_name = coordinator.device_info.get("name")
+        self._attr_unique_id = f"{entry_id}_remote"
```
Unique id is unique per platform per domain, so no need to append `remote` to it unless you plan on adding more remotes

iloveicedgreentea commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+
+            try:
+                # get the mac address from device
+                mac = await self._test_connection(host, port)
```
using self._async_abort_entries_match({CONF_HOST: user_input[CONF_HOST], CONF_PORT: user_input[CONF_PORT]})

because mac may or may not be available (99% probably but not 100) also there would be no reason to configure two of the same hosts anyway

joostlek replied:
There is no reason to configure two of the same, but if people are able to, they can create a mess so we rather try our best to avoid such situatons

iloveicedgreentea replied:
okay makes sense I will grab the mac first, create a unique ID based on that, and abort. Also will use that for enttity unique IDs as you mentioned below

joostlek replied:
Awesome!

iloveicedgreentea committed:
feat: use mac as unique ID for entry

iloveicedgreentea committed:
fix: use unique ID from mac, add proper device

iloveicedgreentea commented:
summary of changes

1. Use mac for unique id, abort config flow if already added
2. Use mac in unique IDs for entities
3. Add proper device which ties together controls and sensors (sensors not included in PR)
4. update library code and error handling
5. removed ping from library, now using python sockets only

iloveicedgreentea committed:
fix: will not be set in init potentially

iloveicedgreentea committed:
fix: access mac

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,120 @@
+            try:
+                # get the mac address from device
+                mac = await self._test_connection(host, port)
+            except CannotConnect:
+                _LOGGER.error("CannotConnect error caught")
+                errors["base"] = "cannot_connect"
+                mac = ""
+
+            if not mac:
+                _LOGGER.error("No MAC address found")
+                if errors.get("base") != "cannot_connect":
+                    errors["base"] = "no_mac"
+            else:
+                _LOGGER.debug("MAC address found: %s", mac)
+                # this will prevent the user from adding the same device twice
+                await self.async_set_unique_id(mac)
+                self._abort_if_unique_id_configured()
+
+            if not errors:
+                _LOGGER.debug("Creating entry")
+                # persist the mac address between HA restarts
+                user_input[CONF_MAC] = mac
+                return self.async_create_entry(
+                    title=DEFAULT_NAME,
+                    data=user_input,
+                )
+        # Whether it's the first attempt or a retry, show the form
```
No need to store the mac again in entry data. Also, What is the difference between `no_mac` and `cannot_connect` for the user? As in, both don't succeed, but they can't do anything special to fix it

iloveicedgreentea replied:
I store the mac there because the integration has to persist it through restarts (the library wont)

iloveicedgreentea replied:
ah I see it lower down, never mind

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,120 @@
+        # connection can raise HeartBeatError if the device is not available or connection does not work
+        except (TimeoutError, aiohttp.ClientError, OSError, HeartBeatError) as err:
+            _LOGGER.error("Error connecting to MadVR: %s", err)
+            raise CannotConnect from err
```
We should not raise exceptions from the library

iloveicedgreentea replied:
made errors locally

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,50 @@
+    ) -> None:
+        """Initialize madvr coordinator."""
+        super().__init__(hass, _LOGGER, name=DOMAIN)
+        self.entry_id = self.config_entry.entry_id
+        # get the mac address from the config entry
```
```suggestion
```

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,50 @@
+        super().__init__(hass, _LOGGER, name=DOMAIN)
+        self.entry_id = self.config_entry.entry_id
+        # get the mac address from the config entry
+        self.mac = self.config_entry.data.get(CONF_MAC)
```
```suggestion
        self.mac = self.config_entry.data.unique_id
        assert self.mac
```

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,89 @@
+        """Return true if the device is on."""
+        return self.madvr_client.is_on
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return the DeviceInfo of this madVR Envy."""
+        return DeviceInfo(
+            identifiers={(DOMAIN, self.coordinator.mac)},
+            name="madVR Envy",
+            manufacturer="madVR",
+            model="Envy",
+        )
```
This can be set in the constructor with `self._attr_device_info =`

iloveicedgreentea committed:
fix: optimize, move error to lib

iloveicedgreentea committed:
feat: add coordinator test, use conf

iloveicedgreentea committed:
fix: use one mock, add init test

iloveicedgreentea committed:
fix: not async

iloveicedgreentea committed:
feat: add remote test

iloveicedgreentea committed:
Merge branch 'dev' into add-madvr-envy

iloveicedgreentea commented:
Updated based on comments and added some more tests

joostlek commented on homeassistant/components/madvr/config_flow.py:
```diff
@@ -0,0 +1,115 @@
+
+    async def _close_test_connection(self, madvr_client: Madvr) -> None:
+        """Close the test connection."""
+        if madvr_client:
```
why the if

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,85 @@
+        self.madvr_client = coordinator.client
+        self._attr_unique_id = coordinator.mac
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, f"{self.coordinator.mac}")},
```
```suggestion
            identifiers={(DOMAIN, coordinator.mac)},
```
Also consider adding the mac as connections. I don't have the right way to do it at me because im on a different PC, but I will come back to this if you haven;t found it in the codebase yet. (just search for connections)

iloveicedgreentea replied:
added connections

joostlek commented on tests/components/madvr/conftest.py:
```diff
@@ -0,0 +1,67 @@
+    return CONFIG_ENTRY
+
+
+@pytest.fixture
+def mock_coordinator(mock_madvr_client) -> Generator[AsyncMock, None, None]:
+    """Mock a MadVRCoordinator."""
+    with patch(
+        "homeassistant.components.madvr.MadVRCoordinator",
+        return_value=AsyncMock(),
+    ) as mock_coordinator:
+        mock_coordinator_instance = mock_coordinator.return_value
+        mock_coordinator_instance.client = mock_madvr_client
+        mock_coordinator_instance.mac = MOCK_MAC
+        yield mock_coordinator_instance
```
We should not mock internals

joostlek commented on tests/components/madvr/conftest.py:
```diff
@@ -0,0 +1,67 @@
+@pytest.fixture
+def mock_config_entry() -> MockConfigEntry:
+    """Mock a config entry."""
+    return CONFIG_ENTRY
```
You can just create the entry here and use the fixture in other places

joostlek commented on tests/components/madvr/conftest.py:
```diff
@@ -0,0 +1,67 @@
+@pytest.fixture
+def mock_madvr_client() -> Generator[AsyncMock, None, None]:
+    """Mock a MadVR client."""
+    with patch(
+        "homeassistant.components.madvr.config_flow.Madvr", autospec=True
+    ) as mock_client:
```
Let's also patch the Madvr in the normal code, check `airgradietn` on how we patch it there

joostlek commented on tests/components/madvr/conftest.py:
```diff
@@ -0,0 +1,67 @@
+        client.power_on = AsyncMock()
+        client.power_off = AsyncMock()
+        client.add_command_to_queue = AsyncMock()
+        with patch("madvr.madvr.send_magic_packet", return_value=True):
+            yield client
```
Not sure why we patch this as this is internal of the library

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```

joostlek replied:
Please also fix the other occurences

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -0,0 +1,90 @@
+    mock_madvr_client.async_cancel_tasks.assert_called_once()
+
+
+async def test_flow_errors(
```
Make sure the test ends in CREATE_ENTRY so we are sure we can recover from an error

joostlek commented on tests/components/madvr/test_coordinator.py:
All these tests test too internal. We should mock the library and what it returns, and we should see the effect of that update on the entities in the state machine. I think testing the data in the coordinator is hard without entities that expose a state, so we should skip that for now, but be sure to implement that in the future

iloveicedgreentea replied:
deleted this test for now

joostlek commented on tests/components/madvr/test_remote.py:
```diff
@@ -0,0 +1,26 @@
+from tests.common import MockConfigEntry
+
+
+async def test_remote_setup(
+    hass: HomeAssistant,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test setup of the remote entity."""
+    await setup_integration(hass, mock_config_entry)
+
+    entity_registry = er.async_get(hass)
+    remote_entity = entity_registry.async_get(f"{REMOTE_DOMAIN}.madvr_envy")
+
+    assert remote_entity is not None
+    assert remote_entity.unique_id == MOCK_MAC
```
You can make this a snapshot test if you like, check `airgradient` for an example. You can then run `pytest ./tests/component/madvr --snapshot-update` and it will generate a .ambr file with the entity fixated

iloveicedgreentea replied:
cool, done

iloveicedgreentea committed:
fix: types

iloveicedgreentea committed:
fix: patch client, expand remote tests

iloveicedgreentea committed:
fix: use snapshot test

iloveicedgreentea commented:
Okay I think I addressed everything

I updated tests, added tests for everything except coordinator, fixed internal patching

iloveicedgreentea committed:
Merge branch 'dev' into add-madvr-envy

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+    await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
+    assert result["title"] == DEFAULT_NAME
+    assert result["data"] == {
+        CONF_HOST: MOCK_CONFIG[CONF_HOST],
+        CONF_PORT: MOCK_CONFIG[CONF_PORT],
+    }
+    assert result["result"].unique_id == MOCK_MAC
```
These can be omitted imo, we already have this tested in the other test

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        result["flow_id"],
+        {CONF_HOST: MOCK_CONFIG[CONF_HOST], CONF_PORT: MOCK_CONFIG[CONF_PORT]},
+    )
+    assert result["type"] == FlowResultType.ABORT
```
```suggestion
    assert result["type"] is FlowResultType.ABORT
```

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+    )
+    await hass.async_block_till_done()
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        result["flow_id"],
+        {CONF_HOST: MOCK_CONFIG[CONF_HOST], CONF_PORT: MOCK_CONFIG[CONF_PORT]},
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -0,0 +1,112 @@
+        result["flow_id"],
+        {CONF_HOST: MOCK_CONFIG[CONF_HOST], CONF_PORT: MOCK_CONFIG[CONF_PORT]},
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek commented on tests/components/madvr/test_remote.py:
```diff
@@ -0,0 +1,95 @@
+        await setup_integration(hass, mock_config_entry)
+
+    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)
+    # await setup_integration(hass, mock_config_entry)
+
+    # entity_registry = er.async_get(hass)
+    # remote_entity = entity_registry.async_get(f"{REMOTE_DOMAIN}.madvr_envy")
+
+    # assert remote_entity is not None
+    # assert remote_entity.unique_id == MOCK_MAC
```
Can be removed

joostlek commented on tests/components/madvr/test_remote.py:
```diff
@@ -0,0 +1,95 @@
+    """Test turning on the remote entity."""
+
+    await setup_integration(hass, mock_config_entry)
+    await hass.async_block_till_done()
```
not needed
```suggestion
```

joostlek commented on tests/components/madvr/test_remote.py:
```diff
@@ -0,0 +1,95 @@
+    await setup_integration(hass, mock_config_entry)
+    await hass.async_block_till_done()
+
+    entity_id = f"{REMOTE_DOMAIN}.madvr_envy"
```
I don't mind if you just used `remote` here instead of the domain

joostlek commented on tests/components/madvr/test_remote.py:
```diff
@@ -0,0 +1,95 @@
+
+    entity_id = f"{REMOTE_DOMAIN}.madvr_envy"
+    remote = hass.states.get(entity_id)
+    assert remote.state == "on"
```
```suggestion
    assert remote.state == STATE_ON
```

joostlek commented on tests/components/madvr/test_remote.py:
```diff
@@ -0,0 +1,95 @@
+    await hass.services.async_call(
+        REMOTE_DOMAIN, SERVICE_TURN_OFF, {ATTR_ENTITY_ID: entity_id}, blocking=True
+    )
+    await hass.async_block_till_done()
```
blocking=True, so can be removed
```suggestion
```

joostlek commented on tests/components/madvr/test_remote.py:
```diff
@@ -0,0 +1,95 @@
+    mock_madvr_client.power_on.assert_called_once()
+
+
+async def test_send_command(
```
Also apply the same things here

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,89 @@
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+if TYPE_CHECKING:
+    from . import MadVRConfigEntry
```
In this case it doesn't have to be in a type checking block

joostlek commented on homeassistant/components/madvr/coordinator.py:
```diff
@@ -0,0 +1,50 @@
+    ) -> None:
+        """Initialize madvr coordinator."""
+        super().__init__(hass, _LOGGER, name=DOMAIN)
+        self.mac = self.config_entry.unique_id
+        assert self.mac
```
```suggestion
        assert self.config_entry.unique_id
        self.mac = self.config_entry.unique_id
```
This way `self.mac` is always `str` so we can avoid the `cast`s in `remote.py`

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,89 @@
+        """Initialize the remote entity."""
+        super().__init__(coordinator)
+        self.madvr_client = coordinator.client
+        self._attr_unique_id = cast(str, coordinator.mac)
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, cast(str, coordinator.mac))},
+            name="madVR Envy",
+            manufacturer="madVR",
+            model="Envy",
+            connections={(CONNECTION_NETWORK_MAC, cast(str, coordinator.mac))},
+        )
```
So please remove the casts after you did that

joostlek commented on homeassistant/components/madvr/remote.py:
```diff
@@ -0,0 +1,89 @@
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        coordinator: MadVRCoordinator,
+        entry_id: str,
```
`hass` and `entry_id` are unused

joostlek commented on homeassistant/components/madvr/strings.json:
```diff
@@ -0,0 +1,21 @@
+    "step": {
+      "user": {
+        "title": "Setup MadVR Envy",
+        "description": "Your device needs to be turned on so we can test the connection. ",
```
```suggestion
        "description": "Your device needs to be turned in order to add the integation. ",
```

joostlek commented on homeassistant/components/madvr/strings.json:
```diff
@@ -0,0 +1,21 @@
+      "user": {
+        "title": "Setup MadVR Envy",
+        "description": "Your device needs to be turned on so we can test the connection. ",
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "port": "[%key:common::config_flow::data::port%]"
+        }
```
Also add a `data_description` part that explains a bit of context, as not everyone knows what a host is. Check other integrations for an example

iloveicedgreentea committed:
fix: update branding

iloveicedgreentea committed:
fix: add description, fix type check

iloveicedgreentea committed:
fix: update tests

iloveicedgreentea committed:
fix: test

joostlek commented on tests/components/madvr/test_config_flow.py:
```diff
@@ -78,13 +76,12 @@ async def test_flow_errors(
-    assert result["type"] == FlowResultType.CREATE_ENTRY
-    assert result["title"] == DEFAULT_NAME
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert result["title"] is DEFAULT_NAME
```
this one should not change

iloveicedgreentea replied:
okay fixed

iloveicedgreentea committed:
fix: update test

iloveicedgreentea committed:
fix: camelcase

joostlek committed:
Fix

iloveicedgreentea committed:
Merge branch 'dev' into add-madvr-envy

joostlek committed:
Merge branch 'refs/heads/add-madvr-envy2' into add-madvr-envy

iloveicedgreentea committed:
feat: strict typing

iloveicedgreentea committed:
fix: strict typing in lib

iloveicedgreentea committed:
fix: type will never be None

iloveicedgreentea committed:
fix: reference to mac, all tests passing

iloveicedgreentea committed:
Merge branch 'dev' into add-madvr-envy
